[Vis Default editor] Fix issue with Rollup

maryia-lapata opened the pull request:
## Summary

1) There is an issue with Rollup Index pattern. When `Date Histogram` agg is selected, the code is broken and `Buckets` group disappeared:

![Aug-01-2019 12-11-06](https://user-images.githubusercontent.com/11524637/62280951-8c440a80-b455-11e9-92bf-9adbc011f4bc.gif)

2) I found one more issue. Rollup Index pattern -> Date Histogram -> interval control should be set in `1m` (default value from Editor Config of Rollup), but it's set in `auto` (due to the [hook](https://github.com/elastic/kibana/pull/42430/files#diff-10db5f379a5d1774e97d202e364f6226L150) is launched before editorConfig is changed)


#### What was done:
- [Reverted](https://github.com/elastic/kibana/commit/ba8a453112aa0d33c063381c5b77cca93200d3d0#diff-d1e2b5c0839b8d02c31d0270350e9387R92) `fieldsFetcher.fetch` to `fieldsFetcher.fetchForWildcard`  for displaying correct `Time Filter field name` during Rollup Index Pattern creation (instead of `timestamp.date_histogram.timestamp`):

     <img height="100" src="https://user-images.githubusercontent.com/11524637/62776892-8339e580-bab4-11e9-8439-c4e0bfa68744.png"/>

- Added a [condition](https://github.com/elastic/kibana/pull/42430/files?file-filters%5B%5D=.tsx#diff-d9ae49afceae6c58c588b850e1a11c22R149) for the effect which sets default params from editor config (it resolves `2`).

- When `getInterval` is invoked with `undefined` value (e.g. when a user cleans the interval field), for Rollup Index Pattern [esInterval](https://github.com/elastic/kibana/blob/master/src/legacy/ui/public/time_buckets/time_buckets.js#L271) is calculated based on original value (`undefined`) rather than default value (`auto`) as for simple Index pattern. It causes an exeption in [parseEsInterval](https://github.com/elastic/kibana/blob/master/src/legacy/core_plugins/data/common/parse_es_interval/parse_es_interval.ts#L56). That's why now `makeLabel` is called safety.



With the changes:

![Aug-09-2019 14-41-54](https://user-images.githubusercontent.com/11524637/62776667-e8d9a200-bab3-11e9-8f42-caab3fc34d59.gif)


### Checklist

Use ~~strikethroughs~~ to remove checklist items you don't feel are applicable to this PR.

- [ ] This was checked for cross-browser compatibility, [including a check against IE11](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#cross-browser-compatibility)
- [ ] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/master/packages/kbn-i18n/README.md)
- [ ] [Documentation](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#writing-documentation) was added for features that require explanation or tutorials
- [ ] [Unit or functional tests](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#cross-browser-compatibility) were updated or added to match the most common scenarios
- [ ] This was checked for [keyboard-only and screenreader accessibility](https://developer.mozilla.org/en-US/docs/Learn/Tools_and_testing/Cross_browser_testing/Accessibility#Accessibility_testing_checklist)

### For maintainers

- [ ] This was checked for breaking API changes and was [labeled appropriately](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#release-notes-process)
- [ ] This includes a feature addition or change that requires a release note and was [labeled appropriately](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#release-notes-process)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/10398/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/10766/)

maryia-lapata commented:
After some debugging I found out that the root cause is deeper. 
What we see: The method `makeLabel` breaks with the error `'field' is required`. 
The reason: The field should be set to a default value when the agg type is changed. It occurs in [`agg_config.js`](https://github.com/elastic/kibana/blob/master/src/legacy/ui/public/vis/agg_config.js#L102). The default value for `Date Histogram` field is `agg.getIndexPattern().timeFieldName` (for Rollup index pattern it equals `timestamp.date_histogram.timestamp`). And then in [deserialize](https://github.com/elastic/kibana/blob/master/src/legacy/ui/public/vis/agg_config.js#L117) method of [`Field` param](https://github.com/elastic/kibana/blob/master/src/legacy/ui/public/agg_types/param_types/field.js#L63) it tries to get a field object by field name (`timestamp.date_histogram.timestamp`), but the index pattern doesn't have a field with such name, it has just `timestamp` field name.

![image](https://user-images.githubusercontent.com/11524637/62448135-c3752d00-b76f-11e9-8823-d309e558555e.png)

This undefined field causes throwing an error. A default value for field params as well as for other params were not set, which in turn produces the error during getting agg description via `makeLabel()` call.

Since Rollup index pattern  has composite `timeFieldName` it seems that it should be a different way for getting a field object by timeFieldName for Rollup index pattern.

@jen-huang if you know could you please explain how `timeFieldName` is formed and why it differs from simple index pattern?

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/11259/)

jen-huang commented:
@maryia-lapata Rollup documents are special in that they are not "real" documents, the field that you are seeing (`timestamp.date_histogram.timestamp`) will eventually be mapped to the original time field when the rollup indices are queried by rollup search.

I am seeing an issue when it comes to creating rollup index pattern:
![image](https://user-images.githubusercontent.com/1965714/62641110-c4d16180-b8f7-11e9-94e3-f3e551c5d963.png)

The fields here should already be `timestamp`, not `timestamp.date_histogram.timestamp`, I think something got lost with typescript conversion. This might be why you are seeing the same field in visualizations.

jen-huang commented:
This is the culprit for the rollup index pattern:
https://github.com/elastic/kibana/commit/ba8a453112aa0d33c063381c5b77cca93200d3d0#diff-d1e2b5c0839b8d02c31d0270350e9387R92

@ppisljar Do you recall why this line was changed from `fieldsFetcher.fetchForWildcard` to `fieldsFetcher.fetch`? Is it possible to change this back?

@maryia-lapata I tested replacing that line with:
```js
indexPattern.fieldsFetcher.fetchForWildcard(pattern, getFetchForWildcardOptions())
```
and the time field is now displaying correctly in visualizations and displaying results:

![image](https://user-images.githubusercontent.com/1965714/62644957-2bf31400-b900-11e9-8eb2-28b377be409d.png)

Could you try adding that change to this fix and see if that unblocks you?

maryia-lapata commented:
@jen-huang thank you so much!

I added the fix for `Time Filter field name` and updated the PR description.

@ppisljar @timroes @jen-huang I'd very appreciate it if you review these changes.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/11698/)

elasticmachine commented:
Pinging @elastic/kibana-app

sulemanof commented on src/legacy/ui/public/vis/editors/default/components/agg.tsx:
```diff
@@ -98,9 +111,6 @@ function DefaultEditorAgg({
   }, [lastParentPipelineAggTitle, isLastBucket, agg.type]);
 
-  // A description of the aggregation, for displaying in the collapsed agg header
-  const aggDescription = agg.type && agg.type.makeLabel ? agg.type.makeLabel(agg) : '';
```
Couldn't we just use `safeMakeLabel` for this purpose?

maryia-lapata replied:
No (without changing of `safeMakeLabel`), because `safeMakeLabel` invokes `makeLabel` of `agg` and doesn't pass any param, but we need to call `makeLabel` of `agg.type` and pass `agg` as param.

sulemanof commented on src/legacy/ui/public/vis/editors/default/components/agg_params.tsx:
```diff
@@ -147,7 +146,7 @@ function DefaultEditorAggParams({
       }
     });
-  }, [agg.type]);
+  }, [agg.type, editorConfig]);
```
`editorConfig` will change if `agg.type` changes:

```suggestion
  }, [editorConfig]);
```

maryia-lapata replied:
Nice catch. 
Fixed.

ppisljar commented:
i think this was changed by mistake from fetchForWildcard to fetch, while updating the method (before it existed on index pattern service and now its on index pattern  and index pattern does not need to be passed in)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/11896/)
